prboom2/src/gl_main.c

```diff
@@ -2328,20 +2328,29 @@ static void gld_DrawSprite(GLSprite *sprite)
 
 static void gld_AddHealthBar(mobj_t* thing, GLSprite *sprite)
 {
-  if (((thing->flags & (MF_COUNTKILL | MF_CORPSE)) == MF_COUNTKILL) && (thing->health > 0))
+  if ((thing->flags & MF_SHOOTABLE) && thing->info->spawnhealth > 0 &&
+      thing->health > 0)
   {
     GLHealthBar hbar;
     int health_percent = thing->health * 100 / thing->info->spawnhealth;
 
-    hbar.cm = -1;
-    if (health_percent <= health_bar_red)
-      hbar.cm = CR_RED;
-    else if (health_percent <= health_bar_yellow)
-      hbar.cm = CR_YELLOW;
-    else if (health_percent <= health_bar_green)
-      hbar.cm = CR_GREEN;
+    hbar.r = hbar.g = -1.0f;
+    if (health_percent <= 0) {
+      hbar.r = 1.0f;
+      hbar.g = 0.0f;
+    } else if (health_percent <= health_bar_red) {
+      hbar.r = 1.0f;
+      hbar.g = (double)health_percent / (health_bar_yellow - health_bar_red);
```

Comment:
Why not cast to `float` here?

Reply:
Oh, need to guard against divide-by-zero here as well. The cutoff values are user-input and are not guarded or bounds-checked. Certain settings would crash the game every time a health bar appeared.

Reply:
done

Reply:
If you managed to set yellow<red in the .cfg file, the block will not be executed because it is always testing for <=red first.